README.md
# cluster-forker

[![CircleCI](https://circleci.com/gh/nivrith/cluster-forker/tree/master.svg?style=svg)](https://circleci.com/gh/nivrith/cluster-forker/tree/master)
[![NPM Downloads](https://img.shields.io/npm/dw/cluster-forker.svg)](https://www.npmjs.com/package/cluster-forker)
[![node](https://img.shields.io/node/v/cluster-forker.svg)](https://www.npmjs.com/package/cluster-forker)
[![License MIT](https://img.shields.io/github/license/nivrith/cluster-forker.svg)](https://github.com/nivrith/cluster-forker/blob/master/LICENSE)

easy node clustering

A single instance of Node.js runs in a single thread. To take advantage of multi-core systems, the user will sometimes want to launch a cluster of Node.js processes to handle the load.

The ClusterForker allows easy creation of child processes that all share server ports.

## Highlights

- Written in Typescript

- Wrapper for the `cluster` module

- Simple API

- Lazy, doesn't actually fork before the `run` method is called

- The worker processes are spawned using the child_process.fork() method, so that they can communicate with the parent via IPC and pass server handles back and forth.

<img src="cluster-forker.gif">

## Installation

npm:

```shell
$ npm install cluster-forker
```

yarn:

```shell
$ yarn add cluster-forker
```

## Usage

> easy node clustering

```js
  //index.js
  const { Cluster } = require('cluster-forker');

  let superCluster = new Cluster();

  let superCluster = new Cluster();

  superCluster
  .master( ( cluster ) => {
    // This block runs only in master worker
    console.log('master is running');
  } )
  .fork( ( cluster ) => {
    http.createServer(function(req, res) {
      res.writeHead(200);
      res.end('process ' + process.pid + ' says hello!');
    }).listen(8000);
  } )
  .on('online', function(worker) {
    console.log('Worker ' + worker.process.pid + ' is online');
  })
  .limit( 4 )
  .forkOnExit((worker, code, signal) => {
    console.log('started new worker fork');
  })
  .onExit( ( worker, code, signal ) => {
    console.log('Worker ' + worker.process.pid + ' died with code: ' + code + ', and signal: ' + signal);
    console.log('Starting a new worker');
  } )
  .run() // Runs the cluster and begins forking

```

```shell

$ node index
> master is running
> Worker 31889 is online
> Worker 31890 is online
> Worker 31891 is online
> Worker 31892 is online

```


## License

MIT © [Nivrith](https://github.com/nivrith)


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### 1.0.0

> 21 July 2019

- :tada: Init [`198d8e2`](https://github.com/nivrith/cluster-forker/commit/198d8e264e427ea32aa486e00f13263e78a270bf)
- Temporarily reduce coverage threshold [`5923554`](https://github.com/nivrith/cluster-forker/commit/59235541e18c29c18ba759f6eb56942a474bfcc6)
- Fix Test [`d034ab2`](https://github.com/nivrith/cluster-forker/commit/d034ab2b55a50ca651314d08cf80a82e995d54a7)


test/index.spec.ts
import {
  Cluster
} from './../src/index';
import {
  expect
} from 'chai';

describe('clusterForker', () => {

  it('Is instance of Cluster', () => {
    expect(new Cluster).to.be.instanceOf(Cluster);
  });

});


src/index.ts
/*!
 * cluster-forker <https://github.com/nivrith/cluster-forker>
 *
 * Copyright (c) Nivrith
 * Licensed under the MIT License.
 */


import cluster, {
  Address,
  ClusterSettings,
  Worker
} from 'cluster';
import {
  cpus
} from 'os';

import net from 'net';
import {
  threadId
} from 'worker_threads';

type ClusterModule = typeof cluster;
type ClusterWorkers = typeof cluster.workers;

/**
 *
 *
 * @export
 * @class Cluster
 */
export class Cluster {
  /**
   *
   *
   * @private
   * @type {ClusterModule}
   * @memberof Cluster
   */
  private cluster: ClusterModule = cluster;

  /**
   *
   *
   * @private
   * @type {(Function | undefined)}
   * @memberof Cluster
   */
  private _master: Function | undefined;

  /**
   *
   *
   * @private
   * @type {(Function | undefined)}
   * @memberof Cluster
   */
  private _forked: Function | undefined;

  /**
   *
   *
   * @static
   * @memberof Cluster
   */
  public static cpus = cpus().length;


  /**
   *
   *
   * @type {number}
   * @memberof Cluster
   */
  public _limit: number = Cluster.cpus;

  /**
   *Creates an instance of Cluster.
   * @memberof Cluster
   */
  constructor() {

  }
  /**
   *
   *
   * @param {( cluster: ClusterModule ) => any} [fn]
   * @returns
   * @memberof Cluster
   */
  public fork( fn ? : ( cluster: ClusterModule ) => any ) {
    this._forked = fn;
    return this;
  }

  /**
   *
   *
   * @returns
   * @memberof Cluster
   */
  public async run() {
    let limit = this._limit;
    // Fork workers.
    if ( this.cluster.isMaster ) {
      for ( let i = 0; i < limit; i++ ) {
        this.cluster.fork();
      }
      return Promise.resolve( {
        worker: this.cluster.worker,
        value: await this.runMaster()
      } )
    } else {
      return Promise.resolve( {
        worker: this.cluster.worker,
        value: await this.runForked()
      } )
    }
  }

  /**
   *
   *
   * @private
   * @returns
   * @memberof Cluster
   */
  private async runMaster() {
    if ( typeof this._master === 'function' ) {
      return await this._master( this.cluster );
    }
  }

  /**
   *
   *
   * @private
   * @returns
   * @memberof Cluster
   */
  private async runForked() {
    if ( typeof this._forked === 'function' ) {
      return await this._forked( this.cluster );
    }
  }

  /**
   *
   *
   * @param {( id: string, worker ? : cluster.Worker ) => any} [callback]
   * @returns
   * @memberof Cluster
   */
  public forEachWorker( callback ? : ( id: string, worker ? : cluster.Worker ) => any ) {
    if ( this.isMaster ) {
      const callbackIsFunction = callback !== undefined && typeof callback === 'function';
      for ( const [ id, worker ] of Object.entries( this.cluster.workers ) ) {
        if ( callbackIsFunction ) {
          callback!( id, worker );
        }
      }
    }
    return this;
  }


  public onExit( listener ? : ( worker: Worker, code: number, signal: cluster.Cluster ) => any ): Cluster {
    if ( listener ) {
      this.cluster.on( 'exit', listener );
    }
    return this;
  }

  public forkOnExit( listener ? : ( worker: Worker, code: number, signal: cluster.Cluster ) => any ): Cluster {
    this.cluster.on( 'exit', ( worker: Worker, code: number, signal: cluster.Cluster ) => {
      this.cluster.fork();
      if ( listener && ( typeof listener === 'function' ) ) {
        listener!( worker, code, signal );
      }
    } )
    return this;
  }

  public removeAllListeners() {
    this.cluster.removeAllListeners();
    return this;
  }

  on( event: string, listener: ( ...args: any[] ) => void ): this;
  on( event: "disconnect", listener: ( worker: Worker ) => void ): this;
  on( event: "exit", listener: ( worker: Worker, code: number, signal: cluster.Cluster ) => void ): this;
  on( event: "fork", listener: ( worker: Worker ) => void ): this;
  on( event: "listening", listener: ( worker: Worker, address: Address ) => void ): this;
  on( event: "message", listener: ( worker: Worker, message: any, handle: net.Socket | net.Server ) => void ): this; // the handle is a net.Socket or net.Server object, or undefined.
  on( event: "online", listener: ( worker: Worker ) => void ): this;
  on( event: "setup", listener: ( settings: ClusterSettings ) => void ): this;
  public on( event: any, listener: any ) {
    this.cluster.on( event, listener )
    return this;
  }
  /**
   *
   *
   * @param {( worker: cluster.Worker ) => void} [listener]
   * @returns
   * @memberof Cluster
   */
  public onfork( listener ? : ( worker: cluster.Worker ) => void ) {
    if ( listener ) {
      this.cluster.on( 'fork', listener );
    }
    return this;
  }


  /**
   *
   *
   * @param {number} [n=Cluster.cpus]
   * @returns
   * @memberof Cluster
   */
  public limit( n: number = Cluster.cpus ) {
    if ( n > Cluster.cpus ) {
      this._limit = Cluster.cpus;
    } else {
      this._limit = n;
    }
    return this;
  }

  /**
   *
   *
   * @param {( cluster: ClusterModule ) => any} [fn]
   * @returns
   * @memberof Cluster
   */
  public master( fn ? : ( cluster: ClusterModule ) => any ) {
    this._master = fn;
    return this;
  }

  /**
   *
   *
   * @readonly
   * @memberof Cluster
   */
  public get isMaster() {
    return this.cluster.isMaster;
  }

  /**
   *
   *
   * @readonly
   * @memberof Cluster
   */
  public get isWorker() {
    return this.cluster.isWorker;
  }

}

export default Cluster;

examples/example.ts
import Cluster from "../src";

let superCluster = new Cluster();

superCluster
  .master( ( cluster ) => {
    // This block runs only in master worker
    console.log( 'Only Master', process.pid );
    // console.log(workers);
    return 'master return';
  } )
  .limit( 25 )// Limit number of child processes . Defaults to max cpu cores
  // If limit is passed greater than num of cpu cores.
  // num of cup cores is used

  .onExit( ( signal, code ) => console.log( 'Exit code: ', process.pid, code ) )
  .fork( ( cluster ) => {
    // runs only in forked workers
    // Do not attempt to call fork on cluster here
    // cluster.fork is undefined here as it can only be called in master process
    return 'fork return';
  } )
  .run() // Runs the cluster and begins forking
  .then( ( result ) => {
    // this block runs in both master and worker before
    // result.value has resolved return value of fork callback

    console.log( result.value, process.pid );
  } );

  // iterate over each worker in master worker
  // Iteration occurs only on master cluster
  superCluster.forEachWorker( ( id, worker ) => {
    console.log( id );
  } )

examples/server.example.ts
import Cluster from "../src";
import http from 'http';

let superCluster = new Cluster();

superCluster
  .master( ( cluster ) => {
    // This block runs only in master worker
    console.log('master is running');
  } )
  .fork( ( cluster ) => {
    http.createServer(function(req, res) {
      res.writeHead(200);
      res.end('process ' + process.pid + ' says hello!');
    }).listen(8000);
  } )
  .on('online', function(worker) {
    console.log('Worker ' + worker.process.pid + ' is online');
  })
  .limit( 4 )
  .forkOnExit((worker, code, signal) => {
    console.log('started new worker fork');
  })
  .onExit( ( worker, code, signal ) => {
    console.log('Worker ' + worker.process.pid + ' died with code: ' + code + ', and signal: ' + signal);
    console.log('Starting a new worker');
  } )
  .run() // Runs the cluster and begins forking
